libfwupdplugin/rustgen.py

```diff
@@ -0,0 +1,449 @@
+
+        # find the type
+        try:
+            val2 = val[split_index:]
+            if val2.endswith("be"):
+                self.endian = Endian.BIG
+                val2 = val2[:-2]
+            elif val2.endswith("le"):
+                self.endian = Endian.LITTLE
+                val2 = val2[:-2]
+            self.type = Type(val2)
+            if self.type == Type.GUID:
+                self.multiplier = 16
+        except ValueError as e:
+            raise ValueError(f"invalid type: {val2}") from e
+
+    def __str__(self) -> str:
+        tmp = f"{self.element_id}: "
+        if self.multiplier:
+            tmp += str(self.multiplier)
+        tmp += str(self.type)
+        if self.endian != Endian.NATIVE:
+            tmp += str(self.endian)
+        if self.default:
+            tmp += f": default={self.default}"
+        elif self.constant:
+            tmp += f": const={self.constant}"
+        elif self.padding:
+            tmp += f": padding={self.padding}"
+        return tmp
+
+
+class Generator:
+    def __init__(self, basename) -> None:
+        self.basename: str = basename
+        self._env = Environment(
+            loader=FileSystemLoader(os.path.dirname(__file__)),
+            autoescape=select_autoescape(),
+            keep_trailing_newline=True,
+        )
+
+    def _process_enums(self, name: str, items: List[EnumItem]) -> Tuple[str, str]:
+
+        # render
+        subst = {
+            "Type": Type,
+            "name": f"Fu{name}",
+            "name_snake": f"fu_{_camel_to_snake(name)}",
+            "items": items,
+        }
+        template_h = self._env.get_template(os.path.basename("fu-rustgen-enum.h.in"))
+        template_c = self._env.get_template(os.path.basename("fu-rustgen-enum.c.in"))
+        return template_c.render(subst), template_h.render(subst)
+
+    def _process_structs(self, name: str, items: List[StructItem]) -> Tuple[str, str]:
+
+        # useful constants
+        size: int = 0
+        for item in items:
+            size += item.size
+        has_constant: bool = False
+        for item in items:
+            if item.constant:
+                has_constant = True
+                break
+
+        # render
+        subst = {
+            "Type": Type,
+            "name": f"Fu{name}",
+            "name_snake": f"fu_struct_{_camel_to_snake(name)}",
+            "items": items,
+            "size": size,
+            "has_constant": has_constant,
+        }
+        template_h = self._env.get_template(os.path.basename("fu-rustgen-struct.h.in"))
+        template_c = self._env.get_template(os.path.basename("fu-rustgen-struct.c.in"))
+        return template_c.render(subst), template_h.render(subst)
+
+    def process_input(self, contents: str) -> Tuple[str, str]:
+        name = None
+        enum_items: List[EnumItem] = []
+        struct_items: List[StructItem] = []
+        offset: int = 0
+
+        # header
+        dst_h = """/* auto-generated, do not modify */
```

Comment:
Why not just put this in the template directly?

Reply:
The template is per-struct or per-enum -- I'd need to add a fu-rustgen.h.in or something -- maybe that makes sense.

Reply:
Done.